File: internal/oauth/ratelimit.go
package oauth

import (
	"net"
	"net/http"
	"strings"
	"sync"
	"time"
)

// loginThrottle is a tiny per-IP failed-login rate limiter. It protects the
// single password gate against brute-force without needing an external
// dependency: any IP that fails N times within failureWindow is blocked for
// blockDuration, and every failure incurs a fixed delay to cap raw attempt
// rate regardless of the counter state.
type loginThrottle struct {
	failureWindow time.Duration
	failureMax    int
	blockDuration time.Duration
	failureDelay  time.Duration

	clock func() time.Time

	mu    sync.Mutex
	fails map[string]*failRecord
}

type failRecord struct {
	count        int
	lastFail     time.Time
	blockedUntil time.Time
}

func newLoginThrottle(clock func() time.Time) *loginThrottle {
	return &loginThrottle{
		failureWindow: loginFailureWindow,
		failureMax:    loginFailureMax,
		blockDuration: loginBlockDuration,
		failureDelay:  loginFailureDelay,
		clock:         clock,
		fails:         map[string]*failRecord{},
	}
}

// check returns (allowed, retryAfter). When allowed is false the caller
// should refuse the attempt and emit a Retry-After hint to the client.
func (t *loginThrottle) check(key string) (bool, time.Duration) {
	now := t.clock()
	t.mu.Lock()
	defer t.mu.Unlock()
	r, ok := t.fails[key]
	if !ok {
		return true, 0
	}
	if now.Before(r.blockedUntil) {
		return false, r.blockedUntil.Sub(now)
	}
	// Block expired or never set, but the failure record is stale —
	// drop it so the next failure starts a fresh window.
	if now.Sub(r.lastFail) > t.failureWindow {
		delete(t.fails, key)
	}
	return true, 0
}

// recordFail bumps the failure counter for the given key. After the Nth
// failure inside the window the IP is blocked for blockDuration.
func (t *loginThrottle) recordFail(key string) {
	now := t.clock()
	t.mu.Lock()
	defer t.mu.Unlock()
	r, ok := t.fails[key]
	if !ok || now.Sub(r.lastFail) > t.failureWindow {
		r = &failRecord{}
		t.fails[key] = r
	}
	r.count++
	r.lastFail = now
	if r.count >= t.failureMax {
		r.blockedUntil = now.Add(t.blockDuration)
	}
}

// recordSuccess clears the failure counter for the given key. Called on a
// successful login so a flapping legitimate user doesn't get blocked by a
// streak of fat-finger attempts they then corrected.
func (t *loginThrottle) recordSuccess(key string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	delete(t.fails, key)
}

// gc is invoked from Server.gc(). Sweeps expired records so the map doesn't
// keep growing for one-off scanners that hit /authorize once and disappear.
func (t *loginThrottle) gc() {
	now := t.clock()
	t.mu.Lock()
	defer t.mu.Unlock()
	for k, r := range t.fails {
		if now.After(r.blockedUntil) && now.Sub(r.lastFail) > t.failureWindow {
			delete(t.fails, k)
		}
	}
}

// clientIP returns a stable per-source identifier for rate-limiting under
// the deployment topology where this process sits behind a single reverse
// proxy on the loopback interface (Apache on the same box).
//
// XFF spoofing was a real bug in the first cut: many proxies, Apache
// included, *append* the real client IP to a client-supplied
// X-Forwarded-For value, so the leftmost element is whatever the attacker
// chose to send. Per-IP throttling that keys on a spoofable string is no
// throttling at all.
//
// We now:
//   - if the immediate connection is NOT loopback, trust RemoteAddr and
//     ignore XFF entirely — nothing we'd believe is upstream.
//   - if it IS loopback, take the *rightmost* XFF entry — the one our
//     trusted reverse proxy just appended — and ignore everything to the
//     left. An attacker can prepend whatever they like; they can't change
//     what the proxy adds.
//
// A wrong/missing header keys on the proxy's own IP, which means heavy
// abuse falls back to a single bucket instead of zero throttling — fine
// for a single-user prototype, and the error mode is loud (legitimate
// users notice immediately).
func clientIP(r *http.Request) string {
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		host = r.RemoteAddr
	}
	ip := net.ParseIP(host)
	if ip == nil || !ip.IsLoopback() {
		return host
	}
	xff := r.Header.Get("X-Forwarded-For")
	if xff == "" {
		return host
	}
	// Use the rightmost element — that's the one our local reverse proxy
	// added. Anything to the left of it is attacker-controlled.
	if i := strings.LastIndex(xff, ","); i >= 0 {
		return strings.TrimSpace(xff[i+1:])
	}
	return strings.TrimSpace(xff)
}

File: src/oauth/ratelimit.rs
//! Per-subject token bucket + per-IP login/DCR window throttles + client-IP
//! extraction. Port of Go's `internal/oauth/ratelimit.go`. All time comes from
//! an injectable clock (unix seconds) so tests are deterministic.

use std::collections::HashMap;
use std::net::IpAddr;

use parking_lot::Mutex;

use super::Clock;

/// Per-subject token bucket for the Bearer-authenticated `/mcp` endpoint. Caps
/// the amplification a leaked JWT can drive against every downstream surface.
pub struct SubLimiter {
    rate: f64,
    burst: f64,
    ttl: i64,
    clock: Clock,
    buckets: Mutex<HashMap<String, Bucket>>,
}

struct Bucket {
    tokens: f64,
    last_refill: i64,
    last_use: i64,
}

impl SubLimiter {
    pub fn new(per_sec: f64, burst: u32, ttl_secs: i64, clock: Clock) -> Self {
        Self {
            rate: per_sec,
            burst: burst as f64,
            ttl: ttl_secs,
            clock,
            buckets: Mutex::new(HashMap::new()),
        }
    }

    /// True if the bucket for `key` has spare capacity. A new key starts with
    /// its full burst, so the first call is never denied.
    pub fn allow(&self, key: &str) -> bool {
        let now = (self.clock)();
        let mut buckets = self.buckets.lock();
        let b = buckets.entry(key.to_string()).or_insert(Bucket {
            tokens: self.burst,
            last_refill: now,
            last_use: now,
        });
        let elapsed = (now - b.last_refill).max(0) as f64;
        b.tokens = (b.tokens + elapsed * self.rate).min(self.burst);
        b.last_refill = now;
        b.last_use = now;
        if b.tokens >= 1.0 {
            b.tokens -= 1.0;
            true
        } else {
            false
        }
    }

    /// Drops buckets idle past the TTL so the map stays bounded under churn.
    pub fn gc(&self) {
        let now = (self.clock)();
        self.buckets
            .lock()
            .retain(|_, b| now - b.last_use <= self.ttl);
    }
}

/// A per-IP failed-attempt rate limiter: N events within `window` per key
/// triggers a block for `block`, and every event optionally incurs a fixed
/// delay. Reused for both the login form and `/oauth/register` (DCR), with
/// different tuning.
pub struct Throttle {
    window: i64,
    max: i32,
    block: i64,
    pub failure_delay_ms: u64,
    clock: Clock,
    fails: Mutex<HashMap<String, FailRecord>>,
}

#[derive(Default)]
struct FailRecord {
    count: i32,
    last_fail: i64,
    blocked_until: i64,
}

impl Throttle {
    pub fn new(window: i64, max: i32, block: i64, failure_delay_ms: u64, clock: Clock) -> Self {
        Self {
            window,
            max,
            block,
            failure_delay_ms,
            clock,
            fails: Mutex::new(HashMap::new()),
        }
    }

    /// Atomic check + count. If not blocked, admits AND increments (arming the
    /// block if the threshold trips); returns `(true, 0)`. If blocked, returns
    /// `(false, retry_secs)` without changing state. Counting on admission is
    /// what closes the parallel-burst race.
    pub fn try_acquire(&self, key: &str) -> (bool, i64) {
        let now = (self.clock)();
        let mut fails = self.fails.lock();
        if let Some(r) = fails.get(key)
            && now < r.blocked_until
        {
            return (false, r.blocked_until - now);
        }
        let r = fails.entry(key.to_string()).or_default();
        if now - r.last_fail > self.window {
            // Stale window — start fresh.
            *r = FailRecord::default();
        }
        r.count += 1;
        r.last_fail = now;
        if r.count >= self.max {
            r.blocked_until = now + self.block;
        }
        (true, 0)
    }

    /// Clears the counter for `key` (a successful login) so a fat-fingering
    /// legitimate user isn't penalized later.
    pub fn record_success(&self, key: &str) {
        self.fails.lock().remove(key);
    }

    pub fn gc(&self) {
        let now = (self.clock)();
        self.fails
            .lock()
            .retain(|_, r| !(now > r.blocked_until && now - r.last_fail > self.window));
    }
}

/// A stable per-source identifier for rate-limiting under the deployment
/// topology where this process sits behind a single reverse proxy on loopback.
///
/// XFF spoofing is real: proxies append the client IP, so the LEFTMOST element
/// is attacker-chosen. We therefore: trust RemoteAddr and ignore XFF when the
/// peer isn't loopback; when it IS loopback, take the RIGHTMOST XFF entry (the
/// one our trusted proxy just appended) and ignore everything left of it.
pub fn client_ip(remote: IpAddr, xff: Option<&str>) -> String {
    if !remote.is_loopback() {
        return remote.to_string();
    }
    match xff {
        None => remote.to_string(),
        Some(xff) => match xff.rsplit_once(',') {
            Some((_, last)) => last.trim().to_string(),
            None => xff.trim().to_string(),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Arc;
    use std::sync::atomic::{AtomicI64, Ordering};

    fn fixed_clock(t: i64) -> (Clock, Arc<AtomicI64>) {
        let now = Arc::new(AtomicI64::new(t));
        let n = now.clone();
        (Arc::new(move || n.load(Ordering::SeqCst)), now)
    }

    #[test]
    fn sub_limiter_burst_then_refill() {
        let (clock, now) = fixed_clock(0);
        let lim = SubLimiter::new(5.0, 3, 1800, clock);
        // burst 3 in the same instant
        assert!(lim.allow("a"));
        assert!(lim.allow("a"));
        assert!(lim.allow("a"));
        assert!(!lim.allow("a"), "burst exhausted");
        // advance 1s → +5 tokens (capped at burst 3)
        now.store(1, Ordering::SeqCst);
        assert!(lim.allow("a"));
        // a different subject has its own full bucket
        assert!(lim.allow("b"));
    }

    #[test]
    fn throttle_blocks_after_max_then_success_resets() {
        let (clock, _now) = fixed_clock(100);
        let t = Throttle::new(60, 3, 900, 0, clock);
        assert!(t.try_acquire("ip").0);
        assert!(t.try_acquire("ip").0);
        let (ok, _) = t.try_acquire("ip"); // 3rd trips the block
        assert!(ok);
        let (ok4, retry) = t.try_acquire("ip");
        assert!(!ok4, "blocked after max");
        assert!(retry > 0);
        // success clears it
        t.record_success("ip");
        assert!(t.try_acquire("ip").0);
    }

    #[test]
    fn client_ip_xff_logic() {
        let loopback: IpAddr = "127.0.0.1".parse().unwrap();
        let public: IpAddr = "203.0.113.7".parse().unwrap();
        // non-loopback peer → ignore XFF entirely
        assert_eq!(client_ip(public, Some("1.2.3.4")), "203.0.113.7");
        // loopback peer → rightmost XFF entry (proxy-appended)
        assert_eq!(client_ip(loopback, Some("1.2.3.4, 5.6.7.8")), "5.6.7.8");
        assert_eq!(client_ip(loopback, Some("9.9.9.9")), "9.9.9.9");
        // loopback peer, no XFF → peer
        assert_eq!(client_ip(loopback, None), "127.0.0.1");
    }
}
